### ISSUE TITLE
Persistent storage for module extension evaluation.

### ISSUE THREAD

**Issue by aignas** [bazelbuild/bazel#24777]
### Description of the feature request:

This has been discussed in [this Slack thread](https://bazelbuild.slack.com/archives/C014RARENH0/p1734683703459389?thread_ts=1734659269.635149&cid=C014RARENH0)
And it is outlined in [this proposal document](https://docs.google.com/document/d/1TjA7-M5njkI1F38IC0pm305S9EOmxcUwaCIvaSmansg/edit?tab=t.0#heading=h.vvlzzcv3akok)

### What underlying problem are you trying to solve with this feature?

I would like to store URL, sha256 values that I can fetch from the PyPI index. That is something that is a by product of the extension evaluation and I would like to reuse that in future evaluations.

**Comment by fmeum** [comment]
This could be as simple as:
1. adding a new `facts` parameter to `module_ctx.extension_metadata` that accepts an arbitrary Starlark value that can be converted to JSON. This is persisted in the lockfile (in the output base lockfile for reproducible extensions, after it has been added in https://github.com/bazelbuild/bazel/issues/24723).
2. if a lockfile entry for the extension (and OS/arch, if dependent on those) exists and its `bzlTransitiveDigest` matches (but not necessarily its usages or recorded inputs hashes), provide that object back to the implementation in `module_ctx.facts`.

Extensions are expected to only store universally true statements in facts, e.g. the immutable foo registry package bar@1.2.3 has download URL X and hash Y. It's essentially a generic `go.sum` file.

@Wyverald @meteorcloudy

**Comment by aignas** [comment]
Added extra information to the description about the usecase. I think @fmeum has described what I need.

**Comment by fmeum** [comment]
There's another potential application for this: `use_repo_rule` could use it to persist the "recommended attributes" that a repo rule implementation can return (e.g. for `sha256`). This would allow repo rules to persist and reuse hashes without a need for users to specify them manually in `MODULE.bazel`.

@malt3, @sluongng and I are thinking about how to make repo rules work better with BwoB. A "remote-friendly" variant of `http_archive` could rely on this capability to persist the names of files in an archive, which could then be fetched and extracted in a build rule while still allowing analysis time access to the archive contents. 

@Wyverald @meteorcloudy What do you think about that? This could resolve most of the "repo rules should be able to depend on build rules" use cases with much lower implementation complexity.

**Comment by meteorcloudy** [comment]
OMG, what a brilliant idea!! I can only see one downside of this approach, it doesn't support clean offline/vendor build very well since the build inherently requires internet access to fetch dependencies (although it's pretty hermetic due to sha256sum check). 
But this is a really a big performance win for those who don't care about offline/vendor build!

**Comment by fmeum** [comment]
> OMG, what a brilliant idea!! I can only see one downside of this approach, it doesn't support clean offline/vendor build very well since the build inherently requires internet access to fetch dependencies (although it's pretty hermetic due to sha256sum check).

You can still get reliable offline builds if you use a `--disk_cache` and have "built" the build rules that require internet connectivity once. But it's true that this isn't integrated with vendor mode at all. I could see that being doable in the future (e.g. we could mark such rules so that `bazel fetch` could pick them out from the transitive closure and run them), but we would first need to gain some real world user adoption.

**Comment by meteorcloudy** [comment]
Also the custom [glob](https://github.com/fmeum/rbe_repos.bzl/blob/main/internal/glob.bzl) looks a bit scary, not sure how robust it would be in long term. ;D

**Comment by fmeum** [comment]
> Also the custom [glob](https://github.com/fmeum/rbe_repos.bzl/blob/main/internal/glob.bzl) looks a bit scary, not sure how robust it would be in long term. ;D

Yep, that hack is not ideal. It's only needed if you want the same BUILD file to "just work" with the remote `http_archive` rule. Build rules will be restricted in how dynamic the BUILD files they generate can be and it's possible that we would limit this further in the beginning. 

However, I actually mostly copied the glob code from https://github.com/bazelbuild/bazel-skylib/blob/main/rules/directory/private/glob.bzl, so we already have to maintain a Starlark reimplementation of Bazel's glob.

**Comment by gzm0** [comment]
Cross-posting from https://bazelbuild.slack.com/archives/C014RARENH0/p1741940591194479 on @fmeum's request.

This is essentially another use-case for this, where we could replace a non-reproducible extension with something better.

<details>
<summary>Original post with use-case</summary>

Hi, I'm looking for input of the bazel community on a bazel module extension pattern I stumbled upon:

**tl;dr**: is it a good or bad idea to use a non-reproducible extension to put integrity values (shas) in the bazel lockfile ([example code](https://github.com/datahouse/bazel_buildlib/blob/62146a5ccffd9a80a4b6368e5992e99c755637bc/buildlib/private/prisma/prisma_extension.bzl#L85-L99))?

The context is external dependencies where no integrity value is readily available (think: "download X from `https://example.com/<version>/<arch>/binary`").

We need this for [prisma engines](https://www.prisma.io/docs/orm/more/under-the-hood/engines) (but the specifics do not matter).

We ended up using a non-reproducible module extension that
1. Downloads the relevant files in the extension (not a repository created by the extension)
1. Discards the file, but keeps the integrity return value of [module_ctx.download](https://bazel.build/rules/lib/builtins/module_ctx#download)
1. Passes these integrity values to repository rules
1. The repository rules then use that integrity value to download the files again (for real).

Because the extension is non-reproducible, the repository attributes, **including the integrity values** get snapshotted into the bazel lock file --> free lockfile mechanism :tada:

The full setup can be seen in action on https://github.com/datahouse/bazel_buildlib
- [Relevant part of the module extension](https://github.com/datahouse/bazel_buildlib/blob/62146a5ccffd9a80a4b6368e5992e99c755637bc/buildlib/private/prisma/prisma_extension.bzl#L85-L99)
- [How it looks in the lockfile](https://github.com/datahouse/bazel_buildlib/blob/62146a5ccffd9a80a4b6368e5992e99c755637bc/buildlib/MODULE.bazel.lock#L166-L189) (the "version" in the example is a git commit sha, but it could also be "1.2.3").

My question is: **Is this a good idea?** [edit: answered with this issue: OK idea, but the solution in this issue is better].

My thoughts:
- From a usage point of view, it feels super nice. With a little dance, we get locking "for free".
- I can't shake the feeling that this is pushing the lockfile mechanism maybe a bit farther than it was intended to.
- If this is a "good" pattern, files like https://github.com/bazel-contrib/rules_nodejs/blob/main/nodejs/private/node_versions.bzl might just "go away" in the future?

Any thoughts or inputs on this are very much appreciated! TY for taking the time to read to here :)

**Comment by thesayyn** [comment]
Chiming in here;  Being able to persist arbitrary but hermetic facts/metadata in module context would be big game changer for most repository rules we have.

**Comment by fmeum** [comment]
Please take a look at https://github.com/bazelbuild/bazel/pull/26198 and check whether it covers your use cases. I haven't written extensive docs yet, but the PR includes a link to a rules_go PR that uses the feature and hopefully explains the intended usage well enough.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 8.5.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/8.5.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=8.5.0rc1. Thanks!
